elixir/apps/domain/priv/repo/migrations/20250402071836_remove_duplicate_groups.exs

```diff
@@ -0,0 +1,43 @@
+defmodule Domain.Repo.Migrations.RemoveDuplicateGroups do
+  use Ecto.Migration
+
+  def change do
+    # Due to a bug we had where we mistakenly returned an empty list for
+    # group API fetches, we ended up deleting all groups for a particular customer.
+    # We need to clean these up and fix the index such that it won't happen again.
+
+    # Step 1: Remove all duplicate deleted groups
+    execute("""
+    DELETE FROM actor_groups
+    WHERE id IN (
+      SELECT a.id
+      FROM actor_groups a
+      INNER JOIN actor_groups b
+      ON a.account_id = b.account_id
+      AND a.provider_id = b.provider_id
+      AND a.provider_identifier = b.provider_identifier
+      WHERE a.deleted_at IS NOT NULL
+      AND b.deleted_at IS NULL
+      AND a.provider_id IS NOT NULL
+      AND a.provider_identifier IS NOT NULL
```

Comment:
Only asking because we're going to be deleting data.  Have you run this on your local dev instance with the local DB setup similar to how things are currently in prod?

Reply:
Also want to note, this query looks correct.  I just always get concerned when deleting data. 😄

Reply:
Yeah let me see. The mock provider should come in handy for testing this locally.

Reply:
Tested successfully:

1. Switched to `main`
2. Created a bunch of groups via the mock adapter. Deleted them by setting `deleted_at`, then upon the next sync it created duplicates.
3. Switched to this branch.
4. Ran the migration (took 10s)
5. Started the server, and only the non-deleted groups exist
6. Sync successfully upserts existing groups and resurrects deleted ones.

It's quite slow though, so I'm investigating if perhaps we're missing an index somewhere.

Reply:
Ok it's slow because:

1. The mock adapter is generating all-new identity and group names while keeping the provider_identifier constant, causing a worst-case upsert scenario
2. We upsert actor_groups one-by-one

I don't think this is necessarily a blocking issue at this point in time. Will continue chipping away at query optimization over time here.